### PR TITLE
runtime: add SingleContainer when obtaining OCI Spec

### PR DESCRIPTION
**Pull request by yaoyinnan**
When creating a cgroup, add a SingleContainer when obtaining the OCI Spec to apply to ctr, podman, etc. If the CgroupPath is not obtained from the container config, the sandbox id is used as the cgroup path.

Fixes: #5240

Signed-off-by: yaoyinnan <yaoyinnan@foxmail.com>

### CODE REVIEW

**Comment by norbjd** [comment]
Thanks! It solves one of our issues:

```shell
nerdctl pull ubuntu

for i in `seq 20`
do
  nerdctl run --rm --runtime io.containerd.run.kata.v2 ubuntu uname -r &
done
```

Extract of output (notice the errors) **before** the PR with kata 2.5.1:

```
5.15.63         
5.15.63
...
FATA[0030] failed to create shim task: open /sys/fs/cgroup/systemd/vc/tasks: no such file or directory: not found
FATA[0031] failed to create shim task: open /sys/fs/cgroup/cpuset/vc/cpuset.mems: no such file or directory: not found
```

**After** these changes, there are no more errors.

File: src/runtime/pkg/resourcecontrol/utils_linux.go
Comment (norbjd) [reply src/runtime/pkg/resourcecontrol/utils_linux.go:71]: Do you think that it is possible that because this method is not thread-safe, we are experiencing also other issues (like `timed out connecting to vsock`), as shown here: https://gist.github.com/norbjd/1773b2930227d89c347f223f1d9b4ff0#30-containers-in-parallel?

Comment (yaoyinnan) [reply src/runtime/pkg/resourcecontrol/utils_linux.go:71]: Thanks for your suggestion, I will follow this to find the cause of the error and fix it.

Comment (yaoyinnan) [reply src/runtime/pkg/resourcecontrol/utils_linux.go]: I still think random strings should be used, because the cgroup path is empty when nerdctl runs, it will cause conflicts between cgroups of multiple sandboxes, which itself is not correct.

Comment (yaoyinnan) [reply src/runtime/pkg/resourcecontrol/utils_linux.go]: We discussed and get the results on Slack. The problem of `timed out connecting to vsock` was caused by running kata in a virtual machine. This problem was avoided on bare metal.

File: src/runtime/pkg/resourcecontrol/utils_linux.go
Comment (liubin) [inline src/runtime/pkg/resourcecontrol/utils_linux.go:48]: Are `AbsolutePath` and `RandStr` need to be package exported names?

**Review by liubin — APPROVED** [review]

File: src/runtime/pkg/resourcecontrol/utils_linux.go
Comment (yaoyinnan) [reply src/runtime/pkg/resourcecontrol/utils_linux.go]: fixed

File: src/runtime/pkg/resourcecontrol/utils_linux_test.go
Comment (bergwolf) [inline src/runtime/pkg/resourcecontrol/utils_linux_test.go:111]: When removing the old code's UT, please add new UTs for the added code as well.

File: src/runtime/pkg/resourcecontrol/cgroups.go
Comment (bergwolf) [inline src/runtime/pkg/resourcecontrol/cgroups.go:140]: While it lowers the chance of a collision, it is still possible to get two identical paths here. I tend to agree with @fengwang666 that we need to make the function thread-safe to fix the root cause.

File: src/runtime/pkg/resourcecontrol/utils_linux_test.go
Comment (yaoyinnan) [reply src/runtime/pkg/resourcecontrol/utils_linux_test.go:111]: `ValidCgroupPathV1()` and `ValidCgroupPathV2()` were merged into `ValidCgroupPath()`.

File: src/runtime/pkg/resourcecontrol/utils_linux.go
Comment (c3d) [inline src/runtime/pkg/resourcecontrol/utils_linux.go:53]: Why do you need to invent a function to generate a random string? Other places do something like:
```
	sandboxId := uuid.Generate().String()
```

Wouldn't that be enough?

**Review by c3d — CHANGES_REQUESTED** [review]
Hi @yaoyinnan. I believe we should use existing UUID-generating code rather than invent a new way to generate random strings.

File: src/runtime/pkg/resourcecontrol/utils_linux.go
Comment (yaoyinnan) [reply src/runtime/pkg/resourcecontrol/utils_linux.go]: Sorry, I don't know a better way, I have replaced it.

File: src/runtime/virtcontainers/sandbox.go
Comment (yaoyinnan) [reply src/runtime/virtcontainers/sandbox.go:694]: Perhaps a more correct way is to report an error?

Comment (bergwolf) [reply src/runtime/virtcontainers/sandbox.go:694]: How did you reproduce it? I wonder if it is because some container annotation is not set properly. We should always be able to find the podsandbox container when creating a new sandbox in the current kata workflow.

Comment (smile-luobin) [reply src/runtime/virtcontainers/sandbox.go:694]: I have the same issue.  
just use `ctr run --runtime "io.containerd.kata.v2"`, the `containerType` will be `singleContainer`, so `cgroupPath` will be `""`. 

why choose to use a generated random string instead of sandbox.id? 
As far as I know: 

- using randStr, the target cgroup is '/sys/fs/cgroup/vc/{randStr}'; 
- using `sandbox.Id`, the target cgroup is '/sys/fs/cgroup/vc/{sandbox.Id}';

Comment (yaoyinnan) [reply src/runtime/virtcontainers/sandbox.go]: @bergwolf Yes, it should be that SingleContainer was missed when judging the ContainerType of the annotation.

Comment (yaoyinnan) [reply src/runtime/virtcontainers/sandbox.go]: @smile-luobin Yes, it is better to directly use the sandbox id as the cgroup path, I modified it.

File: src/runtime/virtcontainers/sandbox.go
Comment (bergwolf) [inline src/runtime/virtcontainers/sandbox.go:692]: We've handled the `SingleContainer` case. Can we just fail here if spec is not found? Otherwise, I wonder how we can handle other spec-related operations.

Comment (bergwolf) [reply src/runtime/virtcontainers/sandbox.go:692]: In fact, there are places we assume `GetPatchedOCISpec` can always find the spec. Can you please change `GetPatchedOCISpec` to default to the first container in the sandbox if no proper annotations are found, and change the callers accordingly? Let's make sure we have such a dependency applied.

Comment (yaoyinnan) [reply src/runtime/virtcontainers/sandbox.go:701]: @bergwolf Sometimes it is not possible to use the first container in the sandbox as the OCI Spec, because some UTs use empty ContainerConfig to create sandboxes, and if they are modified, the default values ​​in most UTs will need to be changed. Here I am modifying to print debug information in the log instead of returning err. Use the sandbox id as the default cgroup path when the cgroupPath cannot be obtained from the OCI Spec.

Comment (bergwolf) [reply src/runtime/virtcontainers/sandbox.go:701]: @yaoyinnan Let's not add meaningless code just for UTs. If UTs need to be fixed, let's fix them.

**Review by c3d — COMMENTED** [review]
Hi @yaoyinnan, overall, this looks like good progress. I somewhat agree with @bergwolf 's comments, so I'll wait for the next iteration for a final approval, but from my PoV, you are on the right track. Thanks!

**Comment by yaoyinnan** [comment]
@bergwolf @fengwang666 @c3d @fidencio  This PR has basically been completed. Can you review and push to approve this PR for me?

File: src/runtime/pkg/resourcecontrol/utils_linux.go
Comment (jepio) [inline src/runtime/pkg/resourcecontrol/utils_linux.go:26]: The code of this function makes no sense because the second argument is `systemdCgroup := IsSystemdCgroup(path)`.

So the two first branches `if IsSystemdCgroup(path)` and `if systemdCgroup` perform the same check with different returns.

Please fix this somehow.

File: src/runtime/pkg/resourcecontrol/utils_linux.go
Comment (jepio) [inline src/runtime/pkg/resourcecontrol/utils_linux.go:28]: For CgroupsV2 this previously always performed a transformation:
```
return filepath.Join("/", path), nil
```
Why has this been removed? I would prefer if the commit description or PR description explained this.

File: src/runtime/pkg/resourcecontrol/utils_linux.go
Comment (yaoyinnan) [reply src/runtime/pkg/resourcecontrol/utils_linux.go]: Moved to #5153 to resolve the issue.

File: src/runtime/pkg/resourcecontrol/utils_linux.go
Comment (yaoyinnan) [reply src/runtime/pkg/resourcecontrol/utils_linux.go]: Moved to #5153 to resolve the issue.

**Comment by yaoyinnan** [comment]
Moved to #8974 for processing

**Pull request closed without merging**
